backend/PolicyAdmin/PolicyConverter.cs

```diff
@@ -0,0 +1,269 @@
+    {
+
+        public static ResourcePolicy ConvertPolicy(XacmlPolicy xacmlPolicy)
+        {
+            ResourcePolicy policy = new ResourcePolicy();
+            policy.Rules = new List<PolicyRule>();
+
+            foreach (XacmlRule xr in xacmlPolicy.Rules)
+            {
+                PolicyRule rule = new PolicyRule();
+                rule.RuleId = xr.RuleId;
+                rule.Description = xr.Description;
+
+                rule.Subject = new List<string>();
+                rule.Actions = new List<string>();
+                rule.Resources = new List<List<string>>();
+
+                foreach (XacmlAnyOf anyOf in xr.Target.AnyOf)
+                {
+                    foreach (XacmlAllOf allOf in anyOf.AllOf)
+                    {
+                        List<string> subject = null;
+                        List<string> action = null;
+                        List<string> resource = null;
+
+                        subject = GetRuleSubjects(allOf, subject);
+
+                        resource = GetRuleResources(allOf, resource);
+
+                        action = GetRuleActions(allOf, action);
+
+                        if (subject != null)
+                        {
+                            rule.Subject.AddRange(subject);
+                        }
+
+                        if (action != null)
+                        {
+                            rule.Actions.AddRange(action);
+                        }
+
+                        if (resource != null)
+                        {
+                            rule.Resources.Add(resource);
+                        }
+                    }
+
+                }
+
+                policy.Rules.Add(rule);
+            }
+
+            GetObligations(xacmlPolicy, policy);
+
+            return policy;
+        }
+
+        private static void GetObligations(XacmlPolicy xacmlPolicy, ResourcePolicy policy)
+        {
+            foreach (XacmlObligationExpression obligationExpression in xacmlPolicy.ObligationExpressions)
+            {
+                foreach (XacmlAttributeAssignmentExpression attributeAssignmentExpression in obligationExpression.AttributeAssignmentExpressions)
+                {
+                    if (attributeAssignmentExpression.Category.AbsoluteUri.Equals(AltinnXacmlConstants.MatchAttributeCategory.MinimumAuthenticationLevel))
+                    {
+                        XacmlAttributeValue astr = attributeAssignmentExpression.Property as XacmlAttributeValue;
+                        policy.RequiredAuthenticationLevelEndUser = astr.Value;
+                    }
+
+                }
+            }
+        }
+
+        private static List<string> GetRuleActions(XacmlAllOf allOf, List<string> action)
+        {
+            foreach (XacmlMatch match in allOf.Matches.Where(m => m.AttributeDesignator.Category.AbsoluteUri.Equals(XacmlConstants.MatchAttributeCategory.Action)))
+            {
+                if (action == null)
```

Comment:
Isn't it unnecessary to take action/resource/subject list as an input property as well as returning it?

Reply:
Totally agree. Used the auto refactor stuff in VS. This was not very smart